### PR TITLE
kvserver: increase `Migrate` application timeout to 1 minute

### PR DESCRIPTION
**Pull request by erikgrinaker**
**kvserver: increase Migrate application timeout to 1 minute**

This increases the timeout when waiting for application of a `Migrate`
command on all range replicas to 1 minute, up from 5 seconds. It also
adds a cluster setting `kv.migration.migrate_application.timeout` to
control this.

When encountering a range that's e.g. undergoing rebalancing, it can
take a long time for a learner replica to receive a snapshot and respond
to this request, which would cause the timeout to trigger. This is
especially likely in clusters with many ranges and frequent rebalancing
activity.

Touches #72931.

Release note (bug fix): The timeout when checking for Raft application
of upgrade migrations has been increased from 5 seconds to 1 minute, and
is now controllable via the cluster setting
`kv.migration.migrate_application_timeout`. This makes migrations much
less likely to fail in clusters with ongoing rebalancing activity during
upgrade migrations.

**migration: add informative log message for sep intents migrate failure**

The separated intents migration has been seen to go into failure loops
in the wild, with a generic "context deadline exceeded" error. This adds
a more informative log entry with additional hints on how to resolve the
problem.

Release note: None

### CODE REVIEW

File: pkg/migration/migrations/separated_intents.go [hunk 1]
```diff
@@ -562,6 +562,18 @@ func postSeparatedIntentsMigration(
...
+				// We add a message here for 21.2 temporarily, since this has been seen
+				// to fail in the wild.
+				log.Warningf(ctx, `Migrate command failed for range %d: %s.
+Command must be applied on all range replicas (not just a quorum). Please make
+sure all ranges are healthy and fully upreplicated. Heavy rebalancing may
+interfere with the migration, consider temporarily disabling rebalancing with
+these cluster settings:
+kv.allocator.load_based_rebalancing=1
+kv.allocator.range_rebalance_threshold=1`,
+					desc.RangeID, err)
```
Comment (miretskiy) [inline pkg/migration/migrations/separated_intents.go:576]: Perhaps return errors.WithHint to include above suggestion?
Also, should you mention the timeout setting you've added?

File: pkg/migration/migrations/separated_intents.go [hunk 2]
```diff
@@ -562,6 +562,18 @@ func postSeparatedIntentsMigration(
 				return err
 			})
 			if err != nil {
+				// TODO(erikgrinaker): This should be moved into a common helper for
+				// all migrations that e.g. manages checkpointing and retries as well.
+				// We add a message here for 21.2 temporarily, since this has been seen
+				// to fail in the wild.
+				log.Warningf(ctx, `Migrate command failed for range %d: %s.
+Command must be applied on all range replicas (not just a quorum). Please make
+sure all ranges are healthy and fully upreplicated. Heavy rebalancing may
```
Comment (miretskiy) [inline pkg/migration/migrations/separated_intents.go:571]: nit: up-replicated?

**Review by miretskiy — COMMENTED** [review]
Just few nits from me, but I'll let other people on this pr to +1...

File: pkg/migration/migrations/separated_intents.go (hunk 1 above)
Comment (erikgrinaker) [reply pkg/migration/migrations/separated_intents.go:576]: >Perhaps return errors.WithHint to include above suggestion?

You mean in addition to the `log.Warningf()`? Will that make it e.g. show up in job summaries and such?

>Also, should you mention the timeout setting you've added?

Might as well. I doubt it'll help if it doesn't succeed within a minute, but who knows.

File: pkg/migration/migrations/separated_intents.go (hunk 2 above)
Comment (erikgrinaker) [reply pkg/migration/migrations/separated_intents.go:571]: We use "upreplication" in official documentation, so may as well stick to that.

**Review by tbg — APPROVED** [review]
:lgtm: 

Can we backport cluster settings? I think we can (the setting will be ignored on "old" nodes) but it's not a common thing we do and it makes me a little nervous. Perhaps we should backport a version that has an env var instead of a cluster setting.

> Reviewed 1 of 2 files at r1, 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72987#-:-Mp6TE2IEyRtrTqIcjvj:b-xhsv0n)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @itsbilal, @miretskiy, and @stevendanna)

**Review by erikgrinaker — COMMENTED** [review]
Valid point, although I believe there is some precedent for it. The support issue that spawned this was a 35-node cluster, and rolling a cluster of that size to change an envvar when it's already struggling with an upgrade migration seems pretty annoying.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72987#-:-Mp6TyYXBIb-NkHwzknt:b9gwptl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @itsbilal, @miretskiy, and @stevendanna)

**Comment by tbg** [comment]
Yeah, there is precedent, see: https://github.com/cockroachdb/cockroach/pull/70960
(Though I don't think anyone thought about it in that PR necessarily).

How do you feel about adding a step that mutates the setting introduced in this PR in https://github.com/cockroachdb/cockroach/blob/f8cfdaeac3587519ffe4a81a3120230993fb8169/pkg/cmd/roachtest/tests/versionupgrade.go#L48-L80

**Comment by erikgrinaker** [comment]
>How do you feel about adding a step that mutates the setting introduced in this PR in

Sure, I can have a look, but it's unclear to me how this could fail? Only nodes that have the new code will check the cluster setting (they're the only ones that know about it), and when not found it will fall back to the default of a 1 minute timeout. Old nodes who don't know about this setting aren't affected by it, and will use the hardcoded 5 second timeout. What am I missing?

**Comment by tbg** [comment]
Just general paranoia about nodes not gracefully reacting to settings they don't understand. Old nodes will not have the setting to interpret the new kvs they receive here:

https://github.com/cockroachdb/cockroach/blob/904807640f6a28aa82c5e63f0b62de2dbf27b3a0/pkg/server/settingsworker.go#L79-L84

but it's fairly obviously resilient, i.e.

https://github.com/cockroachdb/cockroach/blob/904807640f6a28aa82c5e63f0b62de2dbf27b3a0/pkg/server/settingsworker.go#L38-L50

will ignore the error from

https://github.com/cockroachdb/cockroach/blob/904807640f6a28aa82c5e63f0b62de2dbf27b3a0/pkg/settings/updater.go#L74-L83.

So maybe adding this test doesn't really make much sense, unless it leads to a server crash we wouldn't find out that there's a problem anyway.

**Comment by erikgrinaker** [comment]
>but it's fairly obviously resilient, i.e.

Yeah, if they weren't I'd imagine we'd see crashes during rolling upgrades to new major versions.

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72987#-:-Mp7jr8xARvmzh9CKHCt:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @erikgrinaker, @irfansharif, @itsbilal, @miretskiy, and @stevendanna)

___
*[pkg/kv/kvserver/replica_command.go, line 731 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72987#-Mp7jS3y8X5BZIEK9qmw:-Mp7jS3y8X5BZIEK9qmx:b-51gduk) ([raw file](https://github.com/cockroachdb/cockroach/blob/cca99045345cacc38a300dad7d43423cb995c100/pkg/kv/kvserver/replica_command.go#L731)):*
> ```Go
> 
> 		err = contextutil.RunWithTimeout(ctx, "waiting for merge application",
> 			5*time.Second, func(ctx context.Context) error {
> ```

nit: const this 5*time.Second and describe how unprincipled it is and why it might make sense anyway. Classic case of :sparkles: Jordan’s Law of Sticky Magic Constants :sparkles:.

**Comment by erikgrinaker** [comment]
>Just general paranoia about nodes not gracefully reacting to settings they don't understand.

To be safe, I tested a mixed-version cluster of 21.2.0 without and with this PR, and it works as expected. Old nodes do not list the setting under `SHOW ALL CLUSTER SETTINGS`, new nodes do. Changing the setting updates it on new nodes, and logs a warning on old nodes. Upgrading an old node picks up the new setting. Downgrading a node that has seen the setting no longer shows it in `SHOW ALL CLUSTER SETTINGS`, nor can I set it.

**Comment by erikgrinaker** [comment]
TeamCity estimates that `testrace` is going to take another hour. This change is highly unlikely to trigger any `testrace` failures, and since the main test suite passed I'm going to bors this so we can get the backport ready.

TFTRs!

bors r=tbg,ajwerner,miretskiy

**Pull request merged**